Flatten decoration groups

dneto0 opened the pull request:
Replaces OpDecorationGroup and its uses in OpGroupDecorate and OpMemberDecorate with repeated instances of OpDecorate and OpMemberDecorate.

Fixes https://github.com/KhronosGroup/SPIRV-Tools/issues/602

antiagainst commented on test/opt/flatten_decoration_test.cpp:
```diff
@@ -0,0 +1,183 @@
+                             "OpDecorate %group Flat\n"
+                             "OpDecorate %group NoPerspective\n"
+                             "%group = OpDecorationGroup\n"
+                             "OpGroupDecorate %group %hue %value\n"
```
This is not `OpGroupMemberDecorate`?

dneto0 replied:
%hue and %value are scalars, so you use OpGroupDecorate on them.
Use OpGroupMemberDecorate when you want to apply a set of decorations to a member of a struct type.

antiagainst replied:
Ah, sorry, this comment should be in the corresponding code snippet for `OpGroupMemberDecorate`.

antiagainst commented on test/opt/flatten_decoration_test.cpp:
```diff
@@ -0,0 +1,183 @@
+                            // remove the names as well.
+                            {"OpName %group \"group\"\n"
+                             "%group = OpDecorationGroup\n",
+                             ""},
```
`OpDecorationGroup` with `OpDecorate`s for it but no actual usage via `OpGroup{Member|}Decorate`?

dneto0 replied:
Yes, that's the purpose of this test case.  Maybe the comment needs improving?  The only use is by the OpName.

antiagainst replied:
Sorry, was suggesting a new test case like:

```
OpDecorate %group <stuff>
OpDecorate %group <stuff>
%group = OpDecorationGroup
```

dneto0 replied:
This was a good test case.  Helped expose an infinite loop bug.

antiagainst commented on test/opt/flatten_decoration_test.cpp:
```diff
@@ -0,0 +1,183 @@
+                             "OpGroupMemberDecorate %group %Point 1\n",
+                             "OpMemberDecorate %Point 1 Flat\n"
+                             "OpMemberDecorate %Point 1 Offset 16\n"},
+                            // Multiple OpGroupDecorate
```
I mean here actually.

dneto0 replied:
Yes. Good catch.

antiagainst commented on test/opt/flatten_decoration_test.cpp:
```diff
@@ -0,0 +1,183 @@
+
+INSTANTIATE_TEST_CASE_P(OpGroupMemberDecorate, FlattenDecorationTest,
+                        ::testing::ValuesIn(std::vector<FlattenDecorationCase>{
+                            // One OpGroupDecorate
```
Needs to update the comment to `OpGroupMemberDecorate`.

dneto0 commented:
Thanks again.  I'll take a look at the feedback on tests tomorrow.

dneto0 commented:
Addressed feedback. 
Added test cases of non-member and member decorations that are unrelated to group decorations. This found an infinite loop problem.

antiagainst commented on test/opt/flatten_decoration_test.cpp:
```diff
@@ -180,4 +192,43 @@ INSTANTIATE_TEST_CASE_P(OpGroupMemberDecorate, FlattenDecorationTest,
                              "OpMemberDecorate %Point 2 Offset 16\n"},
                         }), );
 
+INSTANTIATE_TEST_CASE_P(UnrelatedDecorations, FlattenDecorationTest,
```
Nice!

dneto0 commented:
Rebased and squashed into one commit.  Keeping approval.
